File: Tree-shaped Deep Forest/Binary Search Tree/Q2. Insert into a Binary Search Tree/Q2_InsertintoaBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {

        // if tree/subtree is empty, create and return new node
        if (root == null) return new TreeNode(val);

        // if value is smaller, insert into left subtree
        if (root.val > val) {
            root.left = insertIntoBST(root.left, val);
        } 
        // otherwise insert into right subtree
        else {
            root.right = insertIntoBST(root.right, val);
        }

        // return current root to maintain tree structure
        return root;
    }
}

File: Tree-shaped Deep Forest/Binary Search Tree/Q2. Insert into a Binary Search Tree/Q2_InsertintoaBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def insertIntoBST(self, root, val):
        """
        :type root: Optional[TreeNode]
        :type val: int
        :rtype: Optional[TreeNode]
        """

        if root is None:
            return TreeNode(val)  # found insertion position → create new node

        # if value is smaller, go to left subtree
        if root.val > val:
            root.left = self.insertIntoBST(root.left, val)

        # otherwise insert in right subtree
        else:
            root.right = self.insertIntoBST(root.right, val)

        return root  # return current root after insertion